src/d3d9/d3d9_device.cpp

```diff
@@ -1083,7 +1083,10 @@ namespace dxvk {
     if (unlikely(iSwapChain != 0))
       return D3DERR_INVALIDCALL;
 
-    return m_implicitSwapchain->GetFrontBufferData(pDestSurface);
+    // In windowed mode, GetFrontBufferData takes a screenshot of the entire screen.
+    // We use the last used swapchain as a workaround.
+    // Total War: Medieval 2 relies on this.
+    return m_mostRecentlyUsedSwapchain->GetFrontBufferData(pDestSurface);
```

Comment:
Potential use-after-free here, we could grab the pointer to a swap chain that's currently being destroyed, swap chain destructor completes, and *then* the code inside `GetFrontBufferData` grabs the device lock.

I think a reasonable fix would be to unconditionally lock the device in `Swapchain::Release` if called on a secondary (i.e. non-implicit) swap chain, reset the most recently used swap chain there if 0, then also lock here and grab a proper reference to the most recently used swap chain before calling `GetFrontBufferData`.

Reply:
Fixed by adding a DeviceLock to Device::GetFrontBufferData. Our DeviceLock is re-entrant, so that works.